main.cc

```diff
@@ -155,8 +156,17 @@ class stop_signal {
     bool stopping() const {
         return _caught;
     }
-    abort_source& as_local_abort_source() { return _abort_sources.local(); }
-    sharded<abort_source>& as_sharded_abort_source() { return _abort_sources; }
+    void check() const {
+        _abort_sources.local().check();
+    }
+    abort_source& as_local_abort_source() {
+        check();
+        return _abort_sources.local();
```

Comment:
What if the signal is sent after as_local_abort_source() is called?

If someone subscribes to an abort source but doesn't check it, the bug is there, not here.

Reply:
Subscriber is not supposed to check it. Subscription usually comes with a callback that's expected to be called when abort is requested.
The point is that if someone subscribes to already aborted source, the callback is not executed. This fix suggest to check if abort source is aborted prior to subscribing and abort the whole thing altogether (but it [doen't](https://github.com/scylladb/scylladb/pull/19641#issuecomment-2216860803) seem to be correct fix)

Reply:
You mean in between the call to `as_local_abort_source()` and `subscribe()`?
I presume that the signal handler callback is scheduled in a non-preemptive way, so it will run only next time we yield.  In this case subscription will succeed, as well the deferred action to stop the `sharded<service_level_controller>`, so I see no problem with that.

As for the responsibility to check if the subscription succeeded or not (meaning abort was already requested), then yes, the burden still lies on the object to check that.